pkg/controller/podgc/gc_controller.go

```diff
@@ -308,7 +316,26 @@ func (o byCreationTimestamp) Less(i, j int) bool {
 	return o[i].CreationTimestamp.Before(&o[j].CreationTimestamp)
 }
 
-func (gcc *PodGCController) deletePod(ctx context.Context, namespace, name string) error {
+func (gcc *PodGCController) failAndDeletePodWithCondition(ctx context.Context, namespace, name string, condition *v1.PodCondition) error {
 	klog.InfoS("PodGC is force deleting Pod", "pod", klog.KRef(namespace, name))
+	if utilfeature.DefaultFeatureGate.Enabled(features.JobPodFailurePolicy) {
+		pod, err := gcc.kubeClient.CoreV1().Pods(namespace).Get(ctx, name, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+		newStatus := pod.Status.DeepCopy()
+		if condition != nil {
+			apipod.UpdatePodCondition(newStatus, condition)
+		}
+		// Mark the pod as failed - this is especially important in case the pod
+		// is orphaned, in which case the pod would remain in the Running phase
+		// forever as there is no kubelet running to change the phase.
+		if pod.Status.Phase != v1.PodSucceeded {
```

Comment:
@smarterclayton you might be interested in this part.

In general, this causes the apiserver to do an immediate deletion of the object. But GC always did that by setting the grace period to zero.